transfer_learning.py

```diff
@@ -0,0 +1,229 @@
+# from torchvision import models
+
+
+class dataset():
+    '''
+    Class defining the creation of a huggingface Dataset for training. Probably doesn't
+    need to be a class rather than just a couple functions, but fuck it.
+    '''
+    def __init__(self, data_path='data/tensors.pt', label_path='data/anime_label_map.json', noise_path=None, train_test_split=0):
+        self.data_path = data_path
+        self.label_path = label_path
+        self.noise_path = noise_path
+        self.train_test_split=train_test_split
+
+    def get_dataset(self):
+        '''
+        Loads a pytorch tensor file at a path and returns a huggingface Dataset for
+        ease of use in fine-tuning a model.
+        '''
+        
+        # load all data, optionally with noise
+        tensors = torch.load(self.data_path)
+        if self.noise_path is not None:
+            print('Warning: adding noise creates a very, very large dataset. Waiting 5 seconds in case you want to cancel')
+            time.sleep(5)
+            noise = torch.load(self.noise_path)
+            tensors += noise
+
+        all_data = Dataset.from_dict({'pixel_values': [t[0] for t in tensors], 'labels': [t[1] for t in tensors]})
+        
+        # optionally apply a train/test split
+        if self.train_test_split > 0:
+            split = all_data.train_test_split(test_size=self.train_test_split, seed=0)
+            return split['train'], split['test']
+        else:
+            return all_data
+
+    def get_labels(self):
+        '''
+        Return a dictionary mapping labels to encodings
+        '''
+        with open(self.label_path, 'r') as f:
+            labels = json.load(f)
+
+        if self.noise_path is None:
+            del labels['11']
+
+        return labels
+
+
+class classifier():
+    '''
+    Class defining classifier model, providing methods for training model and making predictions.
+    '''
+    def __init__(self, model, labels, save_path, learning_rate=5e-5, batch_size=10, num_train_epochs=3, weight_decay=0.1):
+        self.learning_rate = learning_rate
+        self.per_device_train_batch_size = batch_size
+        self.num_train_epochs = num_train_epochs
+        self.weight_decay = weight_decay
+        self.labels = labels
+        self.classifier = AutoModelForImageClassification.from_pretrained(model, num_labels=len(labels), ignore_mismatched_sizes=True)
+        self.save_path = save_path
+
+    def compute_metrics(self, pred):
+        '''
+        Helper function for evaluating models using the Trainer module.
+        '''
+        metric = evaluate.load('accuracy')
+        logits, labels = pred
+        predictions = np.argmax(logits, axis=-1)
+
+        return metric.compute(predictions=predictions, references=labels)
+
+    def train_model(self, train, test):
+        '''
+        Perform transfer learning on a pretrained model and save the model
+        image to local machine.
+        '''
+
+        # set device for model execution
```

Comment:
If I remember correctly we might need to load the tensors differently for gpu usage. But we can get to that when we need it, don't think it needs changing right now.